### PR TITLE
HTML tables have symbols for classifications.

### LINKED ISSUE #39: HTML tables: replace classifications with images
As in the LaTeX tables, we would prefer to use symbols for benchmark classifications, without packaging PNGs with the HTML tables.

### PR DESCRIPTION
**Pull request by snim2**
Fixes #39 

### Example results table

screenshot from 2018-02-14 17-40-34

### Example diff table

screenshot from 2018-02-14 17-40-29

### CODE REVIEW

**Comment by ltratt** [comment]
This looks like a good start. One thing I think we need to fix (don't ask me how!) is to vertically centre the symbols. e.g. at the moment the flat symbol looks like an underscore. I also think the warmup and slowdown symbols might be too tall for the font they're next to? [It might take a bit of thinking to work out if this is solvable. It's above my pay grade!]

**Comment by snim2** [comment]
Hrm. The symbols are exactly as the LaTeX ones, but the HTML is all vertically centered, whereas with LaTeX the bottom of the bounding box is where the horizontal line of a letter "L" would be, I think. This screenshot with bounding boxes might make it clearer:

bounding

**Comment by ltratt** [comment]
So what that shows is that the bottom of the bounding box in HTML is not the same as the base line (whereas I think in LaTeX it is the same as the base line). The nicest fix, I think, would be to make the symbols height be from the base line. [This will solve both issues I noted I suspect.] I don't know how to do this, but I suspect it's possible somehow. Maybe search for something "align box to base line" or similar?

**Comment by snim2** [comment]
This one I think makes more sense:

bounding2

**Comment by ltratt** [comment]
OK, that's looks more like it :) All we need to do now (and I guess you're almost there?) is to make the top of the symbol box be the height of a letter in the font?

**Comment by snim2** [comment]
How about:

bounding3

**Comment by ltratt** [comment]
Something's changed -- the bottom of the box is now below the base line.

**Comment by snim2** [comment]
Right, in that screenshot the bottom of the bbox was the bottom of the text (so where the lowest part of a "g" would be, I guess). The font height is now 20px, same as the symbol heights, so setting the bottom of the bbox to the baseline of the font gives us this:

bounding4

**Comment by ltratt** [comment]
The warmup symbol is still way too tall relative to the font, right? ;) I wonder if you want to use one of the font-based height mechanisms (e.g. `em`)?

**Comment by snim2** [comment]
If you look at the penultimate screenshot, the top and bottom of the bbox was at the lowest point of a "y" and the highest point of an "l", so the font size does match the symbol size. The only other thing we can do is change the height of one or other of them (and changing the height of the canvas means changing all the drawing code, the canvas doesn't scale what's inside it). So, this is making the font size a few pixels bigger than the symbols, but making them both smaller than before so the font doesn't look crazily large :

bounding5

**Comment by ltratt** [comment]
That's looking reasonable enough to me. Shall we try without the boxes drawn?

**Comment by snim2** [comment]
OK, without the boxes some of the lines were only a pixel or two apart and a bit too hard to read, e.g. the good inconsistent looked like one thick line. This is slightly less faithful to the LaTeX but a bit more readable IMO:

screenshot from 2018-02-14 23-11-30
screenshot from 2018-02-14 23-11-27

**Comment by ltratt** [comment]
I think this is pretty much ready for merging.

That said, one thing that your last screenshot made me realise is that we should probably do something to fix the width of the table: having such wide columns makes it harder to read. I don't know if a fixed pixel or `em` width is the best way of doing things (probably the latter?). Maybe @ptersilie has thoughts?


Laurie

**Comment by ptersilie** [comment]
Maybe center the table and don't set it's width to 100% and hopefully that will resize the table to its content?

**Comment by snim2** [comment]
FWIW the width is currently set to 100%, so the sizing in the screenshots is just the width of my browser.

**Comment by ltratt** [comment]
I think Lukas is suggesting *not* setting the width to 100%, because having it stretch to the window width makes it hard to read.

**Comment by snim2** [comment]
When I wrote that last comment on my phone I hadn't seen the one from @ptersilie . Sigh. 

Here's the next one:

screenshot from 2018-02-15 11-07-13
screenshot from 2018-02-15 11-07-02

File: warmup/html.py [hunk 1]
```diff
@@ -115,9 +115,11 @@ def html_histogram(data, nth):
   text-align: left;
 }
 table {
-  width: 100%%;
+  width: 1024px;
```
Comment (ltratt) [inline warmup/html.py]: Can I suggest we make this an `em` or `en` width, so that the table (slightly) adjusts to different font sizes?

**Comment by snim2** [comment]
I'm not totally sure that I understand this. The table does adjust to font sizes if the user zooms in with the browser, so I'm not sure that writing the width in a different unit would do something different? e.g. (175% zoom on my browser):

screenshot from 2018-02-15 11-21-48

**Comment by ltratt** [comment]
When you zoom in, the size of a pixel is also adjusted. But let's say I have a 30pt font size, then the 1024px width will stay the same, so text will start to do odd things.

**Comment by snim2** [comment]
I'm not honestly sure that I can tell the difference on this browser! Anyhow, this is the current look and feel:

screenshot from 2018-02-15 11-25-53
screenshot from 2018-02-15 11-25-52

**Comment by ptersilie** [comment]
Have you tried giving it no width at all? In theory that should resize the table to fit the content.

**Comment by snim2** [comment]
Good idea:

screenshot from 2018-02-15 11-30-15
screenshot from 2018-02-15 11-30-13

**Comment by ltratt** [comment]
OK, we have our solution :) If you squash, I can merge.

**Comment by snim2** [comment]
Squashed!

**Pull request merged**